src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/TaskMethodInvoker.cs

```diff
@@ -74,16 +74,32 @@ public object InvokeEnd(object instance, out object[] outputs, IAsyncResult resu
                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentException(SR.SFxInvalidCallbackIAsyncResult));
             }
 
-            var tuple = invokeTask.Result;
-
-            var task = tuple.Item1 as Task;
-            if (task.IsFaulted)
+            AggregateException ae = null;
+            Tuple<object, object[]> tuple = null;
+            Task task = null;
+            
+            if (invokeTask.IsFaulted)
+            {
+                Fx.Assert(invokeTask.Exception != null, "Task.IsFaulted guarantees non-null exception.");
+                ae = invokeTask.Exception;
+            }
+            else
             {
-                Fx.Assert(task.Exception != null, "Task.IsFaulted guarantees non-null exception.");
+                Fx.Assert(invokeTask.IsCompleted, "Task.Result is expected to be completed");
+                tuple = invokeTask.Result;
+                task = tuple.Item1 as Task;
+                if (task.IsFaulted)
```

Comment:
I checked desktop and it turns out I needed an extra line in the InvokeBegin code to detect a null return value and throw the correct exception there. This should now guarantee a non-null here so I'm adding the assert.